File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/liveride/AdvanceToSegment.java
package net.cyclestreets.liveride;

import net.cyclestreets.routing.Journey;
import net.cyclestreets.routing.Segment;

import org.osmdroid.util.GeoPoint;

final class AdvanceToSegment extends LiveRideState
{
  AdvanceToSegment(final LiveRideState previous,
                   final Journey journey)
  {
    this(previous, journey, journey.segments().get(journey.activeSegmentIndex()+1));
  } // AdvanceToSegment    
  
  AdvanceToSegment(final LiveRideState previous,
                   final Journey journey,
                   final Segment segment) 
  {
    super(previous);
    journey.setActiveSegment(segment);
    notify(segment);
  } // AdvanceToSegment
  
  @Override
  public LiveRideState update(Journey journey, GeoPoint whereIam, int accuracy)
  {
    if(journey.atWaypoint())
      return new PassingWaypoint(this);
    if(journey.atEnd())
      return new Arrivee(this);
    
    return new OnTheMove(this);
  } // update

  @Override
  public boolean isStopped() { return false; }
  @Override
  public boolean arePedalling() { return true; }
}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/liveride/AdvanceToSegment.kt
package net.cyclestreets.liveride

import android.util.Log
import net.cyclestreets.routing.Journey
import net.cyclestreets.routing.Segment
import org.osmdroid.util.GeoPoint

internal class AdvanceToSegment @JvmOverloads constructor(previous: LiveRideState,
                                                          journey: Journey,
                                                          segment: Segment? = journey.segments[journey.activeSegmentIndex() + 1]) : LiveRideState(previous) {
    init {
        journey.setActiveSegment(segment!!)
        // this might not be an important message
        Log.d("importantTest", "AdvanceToSegment Init: ${segment.toString()}")
        notify(segment, true)
    }

    override fun update(journey: Journey, myLocation: GeoPoint, accuracy: Int): LiveRideState {
        if (journey.atWaypoint()) {
            return PassingWaypoint(this)
        }
        if (journey.atEnd()) {
            return Arrivee(this)
        }
        return OnTheMove(this)
    }

    override fun isStopped(): Boolean { return false }
    override fun arePedalling(): Boolean { return true }
}